mautic/mautic
# SugarCRM support for DNC sync and boolean fields

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Yes
| Automated tests included? |
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/270
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) |  https://github.com/mautic/mautic/pull/5498#issuecomment-374266812 & https://github.com/mautic/mautic/issues/6223
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Update based on this comments https://github.com/mautic/mautic/pull/5498#issuecomment-374266812 
In short:
Mautic use for emails 3 DNC status: unsubscribed 1, bounced 2, manual unsubscribed 3
SugarCRM use email_opt_out and invalid_email.
This update add new option to sync DNC status and  try match this fields during fetch/push data in this way:
- email_opt_out  <> unsubscribed, manual unsubscribed
- invalid_email <> bounced

![image](https://user-images.githubusercontent.com/462477/38410760-c05bff36-3985-11e8-91e2-c6d0f365fd1e.png)


**Bug fix:**
Also this PR fix  getFormLeadFields
I noticed if no featuresObject is checked, then no matching fields was  loaded.
Due the BC we need load by default Contacts+Leads matching fealds https://github.com/mautic/mautic/pull/5896/files#diff-63e9186faf376af9f99b582519a697b2R239
Or we can create migration and enable Contacts+ Leads object and no need this change, then just remove getFormLeadFields. Discuss about it here https://github.com/mautic/mautic/pull/5746#discussion_r179689104



[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create SugarCRM integration https://mautic.org/docs/en/plugins/sugarcrm.html and add enable
2. Setup Sugar to Mautic 
2. Go to SugarCRM, edit contact and mark Opted Out | Invalid option next to email address

![image](https://user-images.githubusercontent.com/462477/38410840-f359c3e6-3985-11e8-9b63-9df14b0dff40.png)

4.  run `php app/console mautic:integration:fetchleads --integration=Sugarcrm`
5.  See contact in Mautic, still is contactable. No Opted Out | Invalid wasn't moved to Mautic as DNC (Sugar CRM send via API just Opted Out value)

#### Steps to test this PR:
1. Apply PR and repeat all steps
2. Setup Sugar to Mautic   and enable  **Update emails Do not contact info**
3.  run `php app/console mautic:integration:fetchleads --integration=Sugarcrm`
4. See contact in Mautic, **should marked as unsubsribed**
5. Change Mautic to Sugar CRM integration
6. Update contact and remove DNC 
7.  run `php app/console mautic:integration:fetchleads --integration=Sugarcrm`
8. See SugarCRM and contact Opted Out option - **should disabled**
9. Add DNC bounced status  to contact in Mautic and update contact (re-save)
10. run `php app/console mautic:integration:fetchleads --integration=Sugarcrm`
11.  Contact should **marked as Invalid** 
 Works If both way are ad/remove DNC with success

![image](https://user-images.githubusercontent.com/462477/38521894-3d4c749e-3c47-11e8-849a-eb1eee3381ea.png)
